tests/unit/s2n_session_ticket_test.c

```diff
@@ -793,63 +813,90 @@ int main(int argc, char **argv)
     {
-        EXPECT_NOT_NULL(client_config = s2n_config_new());
-        EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(client_config, 1));
-        EXPECT_SUCCESS(s2n_config_disable_x509_verification(client_config));
-        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
-
-        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
-
-        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
-
-        EXPECT_NOT_NULL(server_config = s2n_config_new());
-        EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(server_config, 1));
-        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
-
-        /* Create nonblocking pipes */
-        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
-
-        /* Set session state lifetime for 15 hours which is equal to the default lifetime of a ticket key */
-        EXPECT_SUCCESS(s2n_config_set_session_state_lifetime(server_config, S2N_SESSION_STATE_CONFIGURABLE_LIFETIME_IN_SECS));
-
-        /* Add one ST key */
-        EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(server_config, ticket_key_name1, s2n_array_len(ticket_key_name1), ticket_key1, s2n_array_len(ticket_key1), 0));
-
-        /* Add second key when the first key is very close to it's encryption peak */
-        uint64_t mock_delay = (server_config->encrypt_decrypt_key_lifetime_in_nanos / 2) - ONE_SEC_IN_NANOS;
-        EXPECT_SUCCESS(s2n_config_set_wall_clock(server_config, mock_nanoseconds_since_epoch, &mock_delay));
-        EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(server_config, ticket_key_name2, s2n_array_len(ticket_key_name2), ticket_key2, s2n_array_len(ticket_key2), 0));
-
-        /* Add third key when the second key is very close to it's encryption peak and
-         * the first key is about to transition from encrypt-decrypt state to decrypt-only state
+        const size_t allowed_failures = 10;
+        size_t failures = 0;
+        bool expected_key_chosen = false;
+
+        /* This test sets up three different ticket encryption keys at various times in their encryption lifetime. The test
+         * is meant to check that the weighted random selection algorithm correctly selects the key that is at its
+         * encryption peak. However the test will sometimes pick a key that is not at its encryption peak because the
+         * selection function uses a weighted random selection algorithm. Here we assert that the likelihood of 
+         * the expected key NOT getting chosen is low. 
+         *
+         * We expect to choose the wrong key 0.02% of the time. This value is drawn from the weight of the expected key, 
+         * which does not change per test run. Therefore, the probability that the test chooses the wrong key
+         * allowed_failures times is 0.0002 ^ 10, which is extremely unlikely to occur.
```

Comment:
0.02% of the time is 1/5000. My understanding is that we're seeing this test fail more often than that.

Letting it fail up to 10 times doesn't seem like the right solution. The logic would definitely be wrong if the test failed more often than it succeeded, but that would pass this test. Like, if we're choosing the wrong key 9/10 times, that is very clearly an error.

Reply:
Probably should mention that we're not seeing this test fail more often than expected. There was a flaw in my testing methodology when I got those results. I can change this to be 0.0002^2, that is still a really small probability, I just wasn't sure what number to choose.
Technically it is possible to choose the wrong key 9/10 times, its just very improbable.

Reply:
I think it's important to identify _what_ is being tested here. If I'm following the assertions correctly, it's wanting to show that when the expected key is selected, it behaves in a certain way. The original test even went as far to assume that it would never pick another one.

Now if we're wanting to test the distribution of it picking another one, we should have a more specialized test for that, as highlighted in #3922; ideally one that doesn't use self-talk but calls the API directly and has a large sample size to eliminate flakiness.

Reply:
The limit of `10` is mostly just an arbitrary number to be able to say that it's very unlikely this will fail in this way again but not have it spin indefinitely.

Reply:
Talked online and lowered allowed failures to 1. Also updated the comment.